src/listener/tcp_listener.rs

```diff
@@ -0,0 +1,116 @@
+use super::is_transient_error;
+
+use crate::listener::Listener;
+use crate::utils::BoxFuture;
+use crate::{log, Server};
+
+use std::fmt::{self, Display, Formatter};
+
+use async_std::net::{self, SocketAddr, TcpStream};
+use async_std::prelude::*;
+use async_std::{io, task};
+
+#[derive(Debug)]
+pub enum TcpListener {
+    FromListener(net::TcpListener),
+    FromAddrs(Vec<SocketAddr>, Option<net::TcpListener>),
+}
+
+impl TcpListener {
+    pub fn from_addrs(addrs: Vec<SocketAddr>) -> Self {
+        Self::FromAddrs(addrs, None)
+    }
+
+    pub fn from_listener(tcp_listener: impl Into<net::TcpListener>) -> Self {
```

Comment:
Everything on listener::TcpListener is effectively pub(crate) because TcpListener itself is only pub(crate). Is there a reason to annotate each method as such as well?